Subject: [PATCH] Skype display name support
Maintainer: Miroslav Koskar (http://mkoskar.com/)
Version: v1

This is a patch for bitlbee-3.2.2 and its skype protocol:

* read DISPLAYNAME which corresponds to the name attached to skype contact
  manually through skype client
* display name is available through ``info`` command
* display name (if specified) takes precedence over full name so
  ``nick_format = %full_name`` returns display name with full name as a fallback
* fix bug which caused only offline contacts to receive correct
  ``%full_name``-based nick names

---
 protocols/skype/skype.c | 49 ++++++++++++++++++++++++++++++++++++++++++-------
 1 file changed, 42 insertions(+), 7 deletions(-)

diff --git a/protocols/skype/skype.c b/protocols/skype/skype.c
index 714babf..d3eafe8 100644
--- a/protocols/skype/skype.c
+++ b/protocols/skype/skype.c
@@ -93,6 +93,7 @@ struct skype_data {
 	int topic_wait;
 	/* These are used by the info command. */
 	char *info_fullname;
+	char *info_displayname;
 	char *info_phonehome;
 	char *info_phoneoffice;
 	char *info_phonemobile;
@@ -117,8 +118,10 @@ struct skype_data {
 	/* Pending user which has to be added to the next group which is
 	 * created. */
 	char *pending_user;
-	/* If the info command was used, to determine what to do with FULLNAME result. */
-	int is_info;
+	/* If the info command was used, to determine what to do with
+	 * FULLNAME & DISPLAYNAME result. */
+	int is_info_fullname;
+	int is_info_displayname;
 };
 
 struct skype_away_state {
@@ -324,6 +327,7 @@ static void skype_parse_users(struct im_connection *ic, char *line)
 	for (i = nicks; *i; i++) {
 		skype_printf(ic, "GET USER %s ONLINESTATUS\n", *i);
 		skype_printf(ic, "GET USER %s FULLNAME\n", *i);
+		skype_printf(ic, "GET USER %s DISPLAYNAME\n", *i);
 	}
 	g_strfreev(nicks);
 }
@@ -380,13 +384,35 @@ static void skype_parse_user(struct im_connection *ic, char *line)
 			imcb_log(ic, "User `%s' changed mood text to `%s'", user, buf);
 	} else if (!strncmp(ptr, "FULLNAME ", 9)) {
 		char *name = ptr + 9;
-		if (sd->is_info) {
-			sd->is_info = FALSE;
+		if (sd->is_info_fullname) {
+			sd->is_info_fullname = FALSE;
 			sd->info_fullname = g_strdup(name);
 		} else {
 			char *buf = g_strdup_printf("%s@skype.com", user);
-			imcb_rename_buddy(ic, buf, name);
-			g_free(buf);
+			if (strlen(name)) {
+				imcb_rename_buddy(ic, buf, name);
+				bee_t *bee = ic->bee;
+				bee_user_t *bu = bee_user_by_handle(bee, ic, buf);
+				if (bu)
+					bee->ui->user_group(bee, bu);
+				g_free(buf);
+			}
+		}
+	} else if (!strncmp(ptr, "DISPLAYNAME ", 12)) {
+		char *name = ptr + 12;
+		if (sd->is_info_displayname) {
+			sd->is_info_displayname = FALSE;
+			sd->info_displayname = g_strdup(name);
+		} else {
+			if (strlen(name)) {
+				char *buf = g_strdup_printf("%s@skype.com", user);
+				imcb_rename_buddy(ic, buf, name);
+				bee_t *bee = ic->bee;
+				bee_user_t *bu = bee_user_by_handle(bee, ic, buf);
+				if (bu)
+					bee->ui->user_group(bee, bu);
+				g_free(buf);
+			}
 		}
 	} else if (!strncmp(ptr, "PHONE_HOME ", 11))
 		sd->info_phonehome = g_strdup(ptr + 11);
@@ -435,6 +461,13 @@ static void skype_parse_user(struct im_connection *ic, char *line)
 			g_free(sd->info_fullname);
 			sd->info_fullname = NULL;
 		}
+		if (sd->info_displayname) {
+			if (strlen(sd->info_displayname))
+				g_string_append_printf(st, "Display Name: %s\n",
+					sd->info_displayname);
+			g_free(sd->info_displayname);
+			sd->info_displayname = NULL;
+		}
 		if (sd->info_phonehome) {
 			if (strlen(sd->info_phonehome))
 				g_string_append_printf(st, "Home Phone: %s\n",
@@ -1527,8 +1560,10 @@ static void skype_get_info(struct im_connection *ic, char *who)
 	ptr = strchr(nick, '@');
 	if (ptr)
 		*ptr = '\0';
-	sd->is_info = TRUE;
+	sd->is_info_fullname = TRUE;
+	sd->is_info_displayname = TRUE;
 	skype_printf(ic, "GET USER %s FULLNAME\n", nick);
+	skype_printf(ic, "GET USER %s DISPLAYNAME\n", nick);
 	skype_printf(ic, "GET USER %s PHONE_HOME\n", nick);
 	skype_printf(ic, "GET USER %s PHONE_OFFICE\n", nick);
 	skype_printf(ic, "GET USER %s PHONE_MOBILE\n", nick);
